src/Share/Survey/SurveyHelper.cs

```diff
@@ -0,0 +1,360 @@
+
+        private void AzPredictorPrompt(SurveyHelper helper, string moduleName, int majorVersion)
+        {
+            helper.Modules[moduleName].LastActiveDate = CurrentDate;
+            helper.Modules[moduleName].ActiveDays += 1;
+        }
+
+        private bool ShouldModulePromptExpire(SurveyHelper helper, string moduleName, int majorVersion)
+            => helper.Modules[moduleName].MajorVersion == majorVersion
+            && (Modules[moduleName].ActiveDays == SurveyTriggerCount + 1 && helper.LastPromptDate == Convert.ToDateTime(Modules[moduleName].LastActiveDate) && helper.Today > helper.LastPromptDate.AddDays(DelayForSecondPrompt)
+                || Modules[moduleName].ActiveDays == SurveyTriggerCount + 2 && helper.LastPromptDate == Convert.ToDateTime(Modules[moduleName].LastActiveDate) && helper.Today > LastPromptDate.AddDays(DelayForThirdPrompt));
+
+        private void ModulePromptExpire(SurveyHelper helper, string moduleName, int majorVersion)
+        {
+            helper.Modules[moduleName].ActiveDays = 0;
+        }
+
+        private void MergeScheduleInfo(ScheduleInfo externalScheduleInfo)
+        {
+            DateTime externalLastPromptDate = Convert.ToDateTime(externalScheduleInfo?.LastPromptDate);
+            IDictionary<string, ModuleInfo> externalModules = new Dictionary<string, ModuleInfo>();
+            foreach(ModuleInfo info in externalScheduleInfo?.Modules)
+            {
+                externalModules[info.Name] = info;
+            }
+
+            HashSet<string> moduleNames = new HashSet<string>(Modules.Keys);
+            moduleNames.UnionWith(new HashSet<string>(externalModules.Keys));
+
+            foreach (string name in moduleNames)
+            {
+                if (externalModules.ContainsKey(name) && (!Modules.ContainsKey(name) || Convert.ToDateTime(Modules[name].LastActiveDate) < Convert.ToDateTime(externalModules[name].LastActiveDate)))
+                {
+                    if (externalLastPromptDate != DateTime.MinValue && Convert.ToDateTime(externalModules[name].LastActiveDate) == externalLastPromptDate)
+                    {
+                        LastPromptDate = externalLastPromptDate;
+                    }
+                    Modules[name] = new ModuleInfo(externalModules[name]);
+                }
+            }
+        }
+
+        private ScheduleInfo GetScheduleInfo()
+        { 
+            return new ScheduleInfo() { LastPromptDate = LastPromptDate.ToString("yyyy-MM-dd"), Modules = Modules.Values.ToList() };
+        }
+
+        private bool ReadFromStream()
+        {
+            StreamReader sr = null;
+            try
+            {
+                if (File.Exists(SurveyScheduleInfoFile))
+                {
+                    sr = new StreamReader(new FileStream(SurveyScheduleInfoFile, FileMode.Open, FileAccess.Read, FileShare.None));                    
```

Comment:
Should we allow multiple read by using `FileShare.Read`? Otherwise once one module is reading the file, the other modules won't be able to read it.

Reply:
You right, read should not block each other.